src/hash.rs

```diff
@@ -0,0 +1,114 @@
+use std::{any::TypeId, fmt::Display, hash::Hash, ops::Deref, sync::Arc};
+
+use dashmap::DashMap;
+
+use crate::CONTAINER_HASH;
+
+struct HashContainer<T: ?Sized> {
+    hashed: DashMap<Arc<T>, ()>,
+}
+
+impl<T: Eq + Hash + ?Sized> HashContainer<T> {
+    pub fn new() -> Self {
+        HashContainer {
+            hashed: DashMap::new(),
+        }
+    }
+}
+
+#[derive(Clone, Debug, PartialEq, Eq, PartialOrd, Ord, Hash)]
+#[repr(transparent)]
+pub struct InternedHash<T: ?Sized + Eq + Hash + 'static>(Arc<T>);
+
+impl<T: ?Sized + Eq + Hash + 'static> InternedHash<T> {
+    pub fn references(this: &Self) -> usize {
+        Arc::strong_count(&this.0)
+    }
+}
+
+impl<T: ?Sized + Eq + Hash + 'static + Display> Display for InternedHash<T> {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        self.0.fmt(f)
+    }
+}
+
+impl<T: ?Sized + Eq + Hash + 'static> Deref for InternedHash<T> {
+    type Target = T;
+
+    fn deref(&self) -> &Self::Target {
+        self.0.deref()
+    }
+}
+
+impl<T: ?Sized + Eq + Hash + 'static> Drop for InternedHash<T> {
+    fn drop(&mut self) {
+        // one reference from this Interned, one from the map
+        if Arc::strong_count(&self.0) == 2 {
+            if let Some(boxed) = CONTAINER_HASH
+                .get_or_init(DashMap::new)
+                .get(&TypeId::of::<T>())
+            {
+                let m = boxed.value().downcast_ref::<HashContainer<T>>().unwrap();
+                m.hashed
+                    .remove_if(&self.0, |k, _v| Arc::strong_count(k) == 2);
+            }
+        }
+    }
+}
+
+/// Intern a shared-ownership reference using hashing (will not clone)
+///
+/// Returns either the given Arc or an already-interned one pointing to an equivalent value.
+pub fn intern_hash_arc<T>(val: Arc<T>) -> InternedHash<T>
+where
+    T: Eq + Hash + Send + Sync + ?Sized + 'static,
+{
+    let type_map = CONTAINER_HASH.get_or_init(DashMap::new);
+
+    // Prefer taking the read lock to reduce contention, only use entry api if necessary.
+    let boxed = if let Some(boxed) = type_map.get(&TypeId::of::<T>()) {
```

Comment:
also interesting to use `TypeId` and `Box<dyn Any>` to have a hashmap for each type. however  since types are known at compile time I wonder if it wouldn't be easier/faster to just use a `OnceCell` per type instead.

Reply:
I copied this pattern from arc-interner, but you’re right, this can probably be done better. Will do in a subsequent PR — just like the Eq optimisation that now has become possible with the custom Interned* types.

Reply:
this is a matter of personal taste, but I'd prefer to do some of the things you're doing here at the call site instead of in a separate crate. one of the reasons being it makes code easier to read. if I don't know what intern-arc is I need to stop reading whatever code I'm reading and understand what intern-arc is doing. I am willing to accept some magical primitives that are not inside the std library like `OnceCell`, or alternatives to std primitives like `parking-lot` or `dashmap` or different channels, because those should be very familiar to someone already familiar with `std`.

Reply:
hmm, turns out that statics cannot be instantiated per type (“cannot use type parameter from outer function”), so in order to keep this thing safe it would need to stay as it is — the other option is to explicitly create interner instances for specific types, but then if you create two of them the equality check would yield false for the same data

Reply:
well, the idea was if you want to do interning for a specific type you do it manually at the call site (instantiating your OnceCell's). as an external crate you can't do that of course